### DISCUSSION TITLE
Execution model differences between top-level and inside of functions

### DISCUSSION THREAD

**Discussion by cmxog** [tact-lang/spec#12] [General]
Currently, `let` and bindings in the top level are not processed in the sequential order of their definition. Because of that, we have two problems:

* The `let` and binding syntax has a different execution model from function bodies where they are strictly sequential. This is rather confusing.
* This discouraged us from performing arbitrary compile-time code execution at the top level as its execution semantics were not clear
* It also bloated the compiler a little bit, having to handle these two environments somewhat differently.

As I recall, the primary motivation for this choice in the compiler was to enable recursive type definitions. The irony is that we don't allow them anyway as we can't really guarantee their finality at this moment. Not without working unions, perhaps.

**Comment by cmxog** [comment]
# Universal execution model

It has been proposed to make it one universal execution model to be used across the code. If we didn't need any recursion, that'd be fine and would greatly simplify things. However, since we will likely need to be able to express something like this, we need to consider options.

## Option 1

Use OCaml-like model (and maybe even a syntax) where mutually dependent functions and types are bundled together with `and` and *maybe* its explicit `rec` notation as well.

Let's make it a bit more material (even if the example is a little bit synthetic)

This should never work as we can't ever initialize it:

```
struct List {
  t: Self
}
```

However, if we assume we have an `Option(T)` union (`Some(T) | None`), perhaps we can find out if there's a way for the type to terminate.

```
struct List {
  t: Option(Self)
}
```

From the looks of it, it doesn't look like we have a particular need to introduce any `rec` notation.

Let's consider a more complicated case.

```
struct Container {
  element: Element,
}
union Element of Container | None
```

Now, they are mutually recursive. We can't define either of them without the other one. If we try to use the `and` syntax

```
struct Container {
  element: Element,
} and
union Element of Container | None
```

or, without a shorthand:

```
let Container = struct { element: Element}
and Element = union of Container | None;
```

then the compiler may regard this as connected statements and, given `and` is *only* used in let-binding context it can use this as an opportunity for type resolution.

**Comment by oleganza** [comment]
Could we maybe avoid any recursive types for v1 entirely? Seems like any variable-length recursive data type has grave issues in terms of gas estimates and DoS risks. Even dictionaries are very limited in use because of that. 

Maybe we can approach this issue with something built around Cells specifically akin to using explicit Boxes in Rust.

**Reply by cmxog** [reply]
https://github.com/tact-lang/tact/issues/24

Either works, I guess. We also have to remember that our type system is designed not only to fit TON VM but the compile-time and the question would be whether we need mutually recursive types in compile-time at the moment.

**Reply by oleganza** [reply]
Good point. I hope we don't need recursive types in compile-time, at least initially.

**Comment by p0lunin** [comment]
Recursive types, especially mutually recursive types, isn't much demand in programming at all if we say about typical imperative programming. There are need of recursive types only in specific cases such as linked lists.

Recursive types and functions are easy to implement in Tact without changing current sequentially computation order. Only question is about mutually recursive types and functions.

Recursive (including mutually recursive) functions can be always transformed to the iterative form. We can never add mutually recursive function to the Tact and ask users to rewrite their programs if they need mutually recursive functions.

Mutually recursive types is more complicated topic. Mutually recursive types can be used but it is gas unefficient. There are solution using our current system to create such types:

```
struct StructBase(T: Type) { val x: T }
enum Enum {
  case StructBase(Self);
  case Nothing;
}
let Struct = StructBase(Enum);
```

Because of we have compile-time functions we can create "partially-created" struct type which is will be defined later in the enum with reference to `Self`.

Such approach can be translated onto the mutually recursive functions:
```
fn test1(data: Data, next: Function) {  
  /* do something with data */
  next(new_data)
}
fn test2(data: Data) {
  /* do something with data */
  test1(new_data, test2)
}
```

So, there are possible solutions of mutually recursive types without adding their explicit support which can be expressed in Tact using other features. I think we must kept sequential computation order.
